standalone/detectors/fasttrack/include/PrefixTreeDepot.h

```diff
@@ -0,0 +1,278 @@
+      std::allocator_traits<LargeAllocator>::destroy(alL, tmp);
+      std::allocator_traits<LargeAllocator>::deallocate(alL, tmp, 1);
+    }
+  }
+};
+
+using Allocator = SelectAllocator<INode>;
+extern ipc::spinlock read_write_lock;
+
+class TreeDepot {
+  INode* _curr_elem = nullptr;
+  Allocator al;
+
+ public:
+  INode* get_current_element() {
+    // std::lock_guard<ipc::spinlock> lg(_read_write_lock);
+    return _curr_elem;
+  }
+
+  void insert_function_element(size_t pc) {
+    std::lock_guard<ipc::spinlock> lg(read_write_lock);
+
+    if (_curr_elem == nullptr) {
+      // the root function has to be called with a big size
+      _curr_elem = al.allocate(5);
+      _curr_elem->parent = nullptr;
+      _curr_elem->pc = pc;
+      return;
+    }
+
+    if (pc == _curr_elem->pc) return;  // done for recursive functions;
+    // no need to use more nodes for same function
+
+    INode* next = _curr_elem->fast_check(pc);
+    if (next) {
+      _curr_elem = next;
+      return;
+    } else {  // it is not the current node or any of the child nodes
+      next = al.allocate(1);
+      next->pc = pc;
+      next->parent = _curr_elem;
+      if (_curr_elem->add_child_node(next, pc)) {
+        _curr_elem = next;
+        return;
+      }
+    }
+    // If we got to here, it means that the current node should be of a
+    // bigger size => allocate next big thing;
+    INode* tmp = al.allocate(_curr_elem->size());
+    *tmp = *_curr_elem;  // might be really slow, use move instead of copy
+                         // assignment operator
+    INode* parent = _curr_elem->parent;
+    if (parent) {
+      parent->change_child_node(tmp, _curr_elem);
+    }
+    // replace so that children of current node point to the new value
+    _curr_elem->change_parent_node(tmp);
+
+    // TODO: MUST replace it too in ThreadState::read_write
+    // Allocator::deallocate(_curr_elem, _curr_elem->size() - 1);
+
+    _curr_elem = tmp;
+    next->parent = _curr_elem;
+
+    // we already know that here we can go to the end.
+    _curr_elem->add_child_node(next, pc);
+    _curr_elem = next;
+  }
+
+  void remove_function_element() {
+    std::lock_guard<ipc::spinlock> lg(read_write_lock);
+
+    if (_curr_elem == nullptr) return;  // func_exit before func_enter
+
+    if (_curr_elem->parent == nullptr) {  // exiting the root function
+      // Allocator::deallocate(_curr_elem, _curr_elem->size() - 1);
+      // not deallocating anymore because node might be used
+      _curr_elem = nullptr;
+      return;
+    }
+    _curr_elem = _curr_elem->parent;
+  }
+
+  std::deque<size_t> make_trace(const std::pair<size_t, INode*>& data) const {
+    std::lock_guard<ipc::spinlock> lg(read_write_lock);
+
+    std::deque<size_t> this_stack;
+    this_stack.emplace_front(data.first);
+
+    INode* iter = data.second;
+    while (iter != nullptr) {
+      this_stack.emplace_front(iter->pc);
+      iter = iter->parent;
+    }
+    return std::move(this_stack);
+  }
+};
+
+#endif
```

Comment:
Comment here header guard
```suggestion
#endif // TREEDEPOT_HEADER_H
```